File: android/app/src/main/java/com/landonkea/makeitso/ActionRouter.kt
// ───────────────────────────────────────────────────────────────────
// ActionRouter.kt — executes actions on Android
// ───────────────────────────────────────────────────────────────────
// After Claude returns action commands, this module routes each
// action to the appropriate handler.
//
// On Android, the available actions are more limited than desktop:
//   - open_app: Open another app on the phone
//   - search_web: Search the internet
//   - send_sms: Send a text message
//   - make_call: Make a phone call
//   - set_alarm: Set an alarm (params: hour, minute, message — all optional;
//     with no params, opens the clock app's "set alarm" screen)
//
// Full system control (like typing, clicking, scrolling) is not
// possible on Android without root access.
// ───────────────────────────────────────────────────────────────────

// This line declares the package (namespace) this file belongs to, matching the folder structure.
package com.landonkea.makeitso

// ── Android system imports ────────────────────────────────────────
// These import Android classes for launching other apps, making calls, and opening URLs.

// Context provides access to Android system services (starting activities, checking package manager, etc.).
import android.content.Context
// Intent is a messaging object used to launch activities, services, or broadcast events (like opening a URL).
import android.content.Intent
// Uri represents a Uniform Resource Identifier — it can be a web URL (https://...), phone number (tel:...), etc.
import android.net.Uri
// AlarmClock provides the well-known Intent actions/extras for asking the device's default
// clock/alarm app to set an alarm, without us needing our own alarm-scheduling code or the
// SET_ALARM permission (ACTION_SET_ALARM is a "broadcast intent" any app can send).
import android.provider.AlarmClock

// "object" creates a singleton — exactly one ActionRouter instance exists for the entire app.
// It acts as a utility that takes an Action object and performs the corresponding Android operation.
object ActionRouter {

    // The main entry point: given an Action and an optional Context, execute the correct behavior.
    // "context" is nullable (Context? = null) because some callers might not provide it.
    // This function's ONLY job is to look at the action type and hand off to the matching
    // single-purpose handler function below — it does no work itself. Splitting each case out
    // into its own named function (executeSearchWeb, executeOpenApp, ...) means each function
    // has exactly one responsibility, which makes each one easy to read and test in isolation.
    fun execute(action: Action, context: Context? = null) {
        // "when" is Kotlin's version of a switch statement — it checks action.type against multiple cases.
        when (action.type) {
            // Each branch below just forwards to a dedicated handler function, passing the
            // context and the params map it needs to do its one job.
            "search_web" -> executeSearchWeb(context, action.params)
            "open_app" -> executeOpenApp(context, action.params)
            "send_sms" -> executeSendSms(context, action.params)
            "make_call" -> executeMakeCall(context, action.params)
            "set_alarm" -> executeSetAlarm(context, action.params)

            // ── Unknown action type ───────────────────────────
            // If Claude returns an action type we don't recognize, log it for debugging.
            else -> {
                // Print the unknown action type to the system console (Logcat) so we can see it during development.
                println("Unknown action type: ${action.type}")
            }
            // End of else/default case.
        }
        // End of when(action.type) block.
    }
    // End of execute() function.

    // ── Search the web ────────────────────────────────────────
    // If Claude says action type is "search_web", open the browser with a DuckDuckGo search.
    // "private" means this helper is only meant to be called from inside ActionRouter.
    private fun executeSearchWeb(context: Context?, params: Map<String, String>) {
        // Get the "query" parameter from the action's params map. "?:" elvis returns if the key is missing.
        val query = params["query"] ?: return
        // Create an Intent that tells Android to open a URL in the browser.
        val intent = Intent(Intent.ACTION_VIEW).apply {
            // Set the URL to a DuckDuckGo search with the user's query embedded in the URL.
            data = Uri.parse("https://duckduckgo.com/?q=$query")
            // FLAG_ACTIVITY_NEW_TASK starts the browser in a new task (a separate window/stack).
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        // Launch the browser by passing the intent to the Android system.
        // "?." means this only runs if context is not null (safe call on nullable variable).
        context?.startActivity(intent)
    }
    // End of executeSearchWeb().

    // ── Open another app ──────────────────────────────────────
    // If Claude says action type is "open_app", launch another app by its package name.
    private fun executeOpenApp(context: Context?, params: Map<String, String>) {
        // Get the "name" parameter (the package name like "com.spotify.music"). Return if missing.
        val packageName = params["name"] ?: return
        // Ask the package manager for a launch Intent for the given package name.
        // getLaunchIntentForPackage returns null if the app is not installed.
        val intent = context?.packageManager?.getLaunchIntentForPackage(packageName)
        // Only try to launch if the intent is not null (the app exists on the device).
        if (intent != null) {
            // Start the other app with the launch intent.
            context.startActivity(intent)
        }
        // If intent is null (app not installed), we silently do nothing.
    }
    // End of executeOpenApp().

    // ── Send an SMS ────────────────────────────────────────────
    // If Claude says action type is "send_sms", open the SMS app with a pre-filled message.
    private fun executeSendSms(context: Context?, params: Map<String, String>) {
        // Get the recipient's phone number from params. Default to empty string if missing.
        val number = params["number"] ?: ""
        // Get the message text from params. Default to empty string if missing.
        val message = params["message"] ?: ""
        // Create an Intent that opens the SMS/messaging app.
        val intent = Intent(Intent.ACTION_VIEW).apply {
            // Set the URI to "sms:<phone number>" so Android knows to use the messaging app.
            data = Uri.parse("sms:$number")
            // Add the message text as an extra ("sms_body" is the standard key for SMS content).
            putExtra("sms_body", message)
            // Start in a new task (separate window stack) so the user can message then return.
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        // Launch the SMS app with the pre-filled number and message.
        context?.startActivity(intent)
    }
    // End of executeSendSms().

    // ── Make a phone call ──────────────────────────────────────
    // If Claude says action type is "make_call", open the phone dialer with a number.
    private fun executeMakeCall(context: Context?, params: Map<String, String>) {
        // Get the phone number from params. "?:" returns early if the number is missing.
        val number = params["number"] ?: return
        // Create an Intent that opens the phone dialer (not a direct call — just the dialer screen).
        val intent = Intent(Intent.ACTION_DIAL).apply {
            // Set the URI to "tel:<phone number>" so the dialer pre-fills this number.
            data = Uri.parse("tel:$number")
            // Start in a new task so the dialer is a separate window from the assistant.
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        // Launch the phone dialer with the pre-filled number.
        context?.startActivity(intent)
    }
    // End of executeMakeCall().

    // ── Set an alarm ───────────────────────────────────────────
    // If Claude says action type is "set_alarm", ask the device's default clock app to set
    // one via the standard AlarmClock.ACTION_SET_ALARM broadcast intent. This does NOT require
    // the SET_ALARM permission in the manifest (any app is allowed to send this intent — it's
    // the clock app itself, not us, that actually schedules the alarm), and it hands the user
    // a normal system alarm-confirmation UI rather than silently scheduling something in the
    // background.
    private fun executeSetAlarm(context: Context?, params: Map<String, String>) {
        // "hour" and "minute" are optional — if Claude doesn't supply them (or supplies a
        // non-numeric value), toIntOrNull() returns null and we fall through to just opening
        // the clock app's own "set alarm" screen instead of crashing on a bad params map.
        val hour = params["hour"]?.toIntOrNull()
        val minute = params["minute"]?.toIntOrNull()
        // Optional label shown on the alarm, e.g. "Wake up".
        val message = params["message"] ?: ""

        val intent = Intent(AlarmClock.ACTION_SET_ALARM).apply {
            if (hour != null) putExtra(AlarmClock.EXTRA_HOUR, hour)
            if (minute != null) putExtra(AlarmClock.EXTRA_MINUTES, minute)
            if (message.isNotEmpty()) putExtra(AlarmClock.EXTRA_MESSAGE, message)
            // SKIP_UI lets the alarm be created without the clock app popping up a
            // confirmation screen first — but only when we actually have a time to set;
            // with no hour/minute at all we WANT the UI so the user can pick a time.
            if (hour != null) putExtra(AlarmClock.EXTRA_SKIP_UI, true)
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        // Only launch if some app on the device can actually handle ACTION_SET_ALARM
        // (resolveActivity returns null otherwise) — avoids a crash on a device/emulator
        // with no clock app installed.
        if (context?.packageManager?.let { intent.resolveActivity(it) } != null) {
            context.startActivity(intent)
        }
    }
    // End of executeSetAlarm().
}
// End of ActionRouter object.

File: ios/MakeItSo/ActionRouter.swift
// ─── ActionRouter.swift ──────────────────────────────────────────────
// This file handles executing the "actions" that Claude returns in its
// response. When Claude says "search the web for weather" or "send a
// text message", it returns an action like "search_web" with parameters.
// This code takes those actions and actually performs them on the iPhone.
//
// On iOS, actions are limited because Apple doesn't let apps control the
// whole system (for security reasons). We can open other apps, search
// the web in Safari, or open the Messages app with pre-filled text, but
// we can't click buttons in other apps or control system settings.
// ──────────────────────────────────────────────────────────────────────

// Import the UIKit framework — this gives us access to the iPhone's app
// management system. UIApplication.shared lets us open URLs (which is
// how we launch Safari, Messages, Phone, and other apps on iOS).
// UIKit is the older iOS framework (SwiftUI is newer) but for opening
// other apps we still need UIKit's capabilities.
import UIKit

// Define the ActionRouter class. This is responsible for figuring out
// what type of action Claude requested and performing the appropriate
// operation on the iPhone. For example, a "search_web" action opens
// Safari with a search query, while a "send_sms" action opens Messages.
class ActionRouter {
    // Create a single shared instance that the whole app uses (singleton
    // pattern). `static` means this belongs to the class itself, not to
    // a specific instance. `shared` is the conventional name so other
    // code can access it as `ActionRouter.shared` from anywhere.
    static let shared = ActionRouter()

    // The main function that executes an action. It takes a ClaudeAction
    // object (which has an actionType string and a params dictionary)
    // and performs the appropriate iOS operation. This is called by
    // ContentView after Claude returns actions in its response.
    func execute(_ action: ClaudeAction) {
        // Use a switch statement to pick the right code based on the
        // action type. A switch is cleaner than multiple if-else chains
        // and makes it easy to add new action types later. Swift's switch
        // must be exhaustive — we handle every case including unknown ones.
        switch action.actionType {

        // Handle the "search_web" action — opens Safari to search for
        // something using DuckDuckGo (a privacy-focused search engine).
        case "search_web":
            // Try to get the search query from the parameters dictionary.
            // The `guard` statement checks if "query" exists in params.
            // If not, we exit the function early (return) because we
            // can't search without something to search for.
            guard let query = action.params["query"] else { return }
            // Convert the query string into a format safe for URLs.
            // Characters like spaces, ?, and & have special meanings in
            // URLs, so we need to "percent-encode" them. For example,
            // a space becomes "%20". The `?? query` means if encoding
            // fails for some reason, we use the original query as-is.
            let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
            // Build a full URL string for DuckDuckGo search. We embed
            // the encoded query into the URL so Safari knows what to
            // search for. The `\()` syntax inserts the variable value
            // directly into the string (string interpolation).
            if let url = URL(string: "https://duckduckgo.com/?q=\(encodedQuery)") {
                // Open the URL using the system's default browser (Safari).
                // UIApplication.shared is the singleton that represents
                // the running app. `.open(url)` tells iOS to open the
                // URL, which launches Safari (or shows a browser tab).
                UIApplication.shared.open(url)
            }
            // Close the if block — if the URL was invalid (which
            // shouldn't happen since we constructed it carefully),
            // we just silently do nothing.

        // Handle the "open_app" action — tries to open another app on
        // the iPhone by using its URL scheme (like "twitter://").
        case "open_app":
            // Get the app name from the parameters. The guard ensures
            // we have a name to work with; if not, we return early
            // because we can't open an unnamed app.
            guard let appName = action.params["name"] else { return }
            // Convert the app name to a URL scheme by adding "://" at
            // the end and making it lowercase. App URL schemes are
            // case-sensitive and usually lowercase (e.g., "twitter",
            // "music", "tel"). The scheme is like a protocol that the
            // other app registers with iOS.
            let scheme = "\(appName.lowercased())://"
            // Try to create a URL from the scheme string. If the scheme
            // is valid (e.g., "twitter://"), we get a URL object. Then
            // check if iOS can open this URL — `canOpenURL` returns true
            // if an app on the device has registered this URL scheme.
            if let url = URL(string: scheme),
               UIApplication.shared.canOpenURL(url) {
                // Open the app. This tells iOS to launch the app that
                // registered this URL scheme. For example, opening
                // "twitter://" would launch the Twitter app if installed.
                UIApplication.shared.open(url)
            }
            // Close the if block — if no app is registered for this
            // scheme, we silently do nothing rather than showing an error.

        // Handle the "send_sms" action — opens the Messages app with
        // a pre-filled phone number and message text.
        case "send_sms":
            // Get both the phone number and message text from the
            // parameters. The guard checks both exist; if either is
            // missing, we return early because we need both to send SMS.
            guard let number = action.params["number"],
                  let message = action.params["message"] else { return }
            // Percent-encode the message text so it's safe for a URL.
            // Special characters in the message (like & or +) need to
            // be encoded so they don't break the URL format.
            let encoded = message.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? message
            // Build a special URL that opens Messages with a pre-filled
            // recipient and message. The "sms:" scheme tells iOS to open
            // Messages, and the "&body=" parameter pre-fills the text.
            // iOS recognizes this format and opens the Messages compose screen.
            if let url = URL(string: "sms:\(number)&body=\(encoded)") {
                // Open the Messages app with the composed message.
                // The user can then review and tap Send to actually
                // send it. We don't send automatically for safety.
                UIApplication.shared.open(url)
            }
            // Close the if block — if the URL is malformed, silently
            // do nothing instead of crashing.

        // Handle the "make_call" action — opens the Phone app with a
        // number ready to dial (but doesn't call automatically).
        case "make_call":
            // Get the phone number from the parameters. Guard ensures
            // we have a number; if not, we return early because we
            // can't call without a number to dial.
            guard let number = action.params["number"] else { return }
            // Build a "tel://" URL with the phone number. This is a
            // special URL scheme that iOS recognizes as a phone call
            // request. The "tel" stands for telephone. The number is
            // used as-is (should already be digits and dashes).
            if let url = URL(string: "tel://\(number)") {
                // Open the Phone app. On a real iPhone, this shows a
                // confirmation dialog asking if the user wants to call
                // the number. On an iPod or iPad without cellular, this
                // does nothing (no phone app available).
                UIApplication.shared.open(url)
            }
            // Close the if block — if the URL is invalid, silently
            // skip the call attempt.

        // Handle ANY action type that we don't recognize. The `default`
        // case catches everything that doesn't match the above cases.
        // This is required because Swift's switch must be exhaustive.
        default:
            // Print a warning to the debug console so developers know
            // Claude returned an action type that we don't handle yet.
            // This helps during development when adding new action types.
            print("Unknown action type: \(action.actionType)")
            // Close the default case.
        }
        // Close the switch statement.
    }
    // Close the execute function.
}
// Close the ActionRouter class.
